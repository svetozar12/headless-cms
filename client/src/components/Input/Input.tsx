import { MenuItem, Select } from "@mui/material";
import { ForwardedRef, forwardRef, ReactNode } from "react";
import Switch from "./subcomponents/Switch";

export interface IInputProps {
  name: string;
  type: string;
  onChange: (e: any) => void;
  value?: any;
  label?: string;
  placeholder?: string;
  extraProps?: React.DetailedHTMLProps<
    React.InputHTMLAttributes<HTMLInputElement>,
    HTMLInputElement
  >;
}

const Input = forwardRef(
  (props: IInputProps, ref: ForwardedRef<HTMLInputElement>) => {
    const {
      name,
      label,
      type,
      placeholder,
      value,
      onChange,
      extraProps,
    } = props;
    let content: ReactNode;
    if (type === "dropdown")
      content = (
        <Select value={value} label="Age" onChange={onChange}>
          <MenuItem value={10}>Ten</MenuItem>
          <MenuItem value={20}>Twenty</MenuItem>
          <MenuItem value={30}>Thirty</MenuItem>
        </Select>
      );
    if (type === "checkbox") content = <Switch ref={ref} />;
    else
      content = (
        <div className="m-2  rounded-md bg-inputBlack p-1">
          <input
            value={value}
            autoComplete={value}
            name={name}
            type={type}
            placeholder={placeholder}
            onChange={onChange}
            className={
              "w-full bg-transparent p-2 text-white autofill:bg-transparent active:border-0"
            }
            {...extraProps}
          />
        </div>
      );
    return (
      <>
        {label && (
          <label htmlFor={name} className="px-2 text-white">
            {label}
          </label>
        )}
        {content}
      </>
    );
  },
);

Input.displayName = "Input";

export default Input;
