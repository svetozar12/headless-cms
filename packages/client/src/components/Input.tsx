export interface IInputProps {
  name: string;
  type: string;
  onChange: (e: any) => void;
  value?: string;
  label?: string;
  placeholder?: string;
  extraProps?: React.DetailedHTMLProps<
    React.InputHTMLAttributes<HTMLInputElement>,
    HTMLInputElement
  >;
}

const Input = (props: IInputProps) => {
  const { name, label, type, placeholder, value, onChange } = props;
  return (
    <>
      {label && (
        <label htmlFor={name} className="px-2 text-white">
          {label}
        </label>
      )}
      <div className="m-2  rounded-md bg-inputBlack p-1">
        <input
          value={value}
          autoComplete={value}
          name={name}
          type={type}
          placeholder={placeholder}
          onChange={onChange}
          className={
            "w-full bg-transparent p-2 text-white autofill:bg-transparent active:border-0"
          }
        />
      </div>
    </>
  );
};

export default Input;
