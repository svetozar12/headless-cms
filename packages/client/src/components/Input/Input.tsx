import { ReactNode } from "react";
import Switch from "./subcomponents/Switch";

export interface IInputProps {
  name: string;
  type: string;
  onChange: (e: any) => void;
  value?: any;
  label?: string;
  placeholder?: string;
  extraProps?: React.DetailedHTMLProps<
    React.InputHTMLAttributes<HTMLInputElement>,
    HTMLInputElement
  >;
}

const Input = (props: IInputProps) => {
  const { name, label, type, placeholder, value, onChange, extraProps } = props;
  let content: ReactNode;
  if (type === "checkbox") content = <Switch innerref={extraProps?.ref} />;
  else
    content = (
      <div className="m-2  rounded-md bg-inputBlack p-1">
        <input
          value={value}
          autoComplete={value}
          name={name}
          type={type}
          placeholder={placeholder}
          onChange={onChange}
          className={
            "w-full bg-transparent p-2 text-white autofill:bg-transparent active:border-0"
          }
          {...extraProps}
        />
      </div>
    );
  return (
    <>
      {label && (
        <label htmlFor={name} className="px-2 text-white">
          {label}
        </label>
      )}
      {content}
    </>
  );
};

export default Input;
