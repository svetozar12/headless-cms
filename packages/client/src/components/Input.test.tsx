import type { ReactWrapper } from "enzyme";
import { mount } from "enzyme";
import { beforeEach, describe, expect, it, vi } from "vitest";
import Input from "./Input";

describe("Input component", () => {
  let wrapper: ReactWrapper;
  const mockedOnChangeEvent = vi.fn();
  const inputTypes = ["text", "password", "email", "number"];
  beforeEach(() => {
    mockedOnChangeEvent.mockReset();
    wrapper = mount(
      <Input
        type="text"
        placeholder="text"
        value="hi"
        onChange={mockedOnChangeEvent}
        name="name"
      />
    );
  });

  it("should render input", () => {
    expect(wrapper.find("input").length).toBe(1);
    expect(wrapper.find("input").prop("type")).toBe("text");
    expect(wrapper.find("input").prop("placeholder")).toBe("text");
    expect(wrapper.find("input").prop("value")).toBe("hi");
    expect(wrapper.find("input").prop("name")).toBe("name");
    expect(wrapper).toMatchSnapshot();
  });
  it("should call onChange", () => {
    wrapper.find("input").simulate("change");
    expect(mockedOnChangeEvent).toHaveBeenCalledTimes(1);
  });
  inputTypes.forEach((type) => {
    it(`should render ${type}`, () => {
      wrapper.setProps({ type });
      expect(wrapper.find("input").prop("type")).toBe(type);
    });
  });
});
