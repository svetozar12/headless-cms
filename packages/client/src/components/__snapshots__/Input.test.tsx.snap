// Vitest Snapshot v1

exports[`Button component > should render button 1`] = `ReactWrapper {}`;

exports[`Button component > should render input 1`] = `ReactWrapper {}`;

exports[`Input component > should render input 1`] = `ReactWrapper {}`;
